### PR TITLE
[API] Milestone endpoints accept names too

### PR DESCRIPTION
**Pull request by 6543**
similar to #12366 but for milestones

is very usefull for https://gitea.com/gitea/go-sdk/pulls/388 ... and so on

### CODE REVIEW

**Comment by 6543** [comment]
as per @lafriks lable `kind/breanking` needed

File: routers/api/v1/repo/milestone.go [hunk 1]
```diff
@@ -249,17 +239,49 @@ func DeleteMilestone(ctx *context.APIContext) {
...
+
+	if err := models.DeleteMilestoneByRepoID(ctx.Repo.Repository.ID, m.ID); err != nil {
 		ctx.Error(http.StatusInternalServerError, "DeleteMilestoneByRepoID", err)
 		return
 	}
 	ctx.Status(http.StatusNoContent)
 }
+
+// getMile get milestone by ID and if not available by name
+func getMile(ctx *context.APIContext) *models.Milestone {
```
Comment (zeripath) [inline routers/api/v1/repo/milestone.go]: I think you should call this getMilestoneByIDOrName and drop the reliance on ctx.

**Comment by 6543** [comment]
@zeripath renamed
But If I remove ctx from this function, I'll have to handle Errors in a seperat function (if dont like to create duplicated code) witch relay on ctx then

**Review by zeripath — APPROVED** [review]

**Comment by CirnoT** [comment]
I really wonder if this is necessary to implement; the API does not need to act via name for user convenience because it's not designed for that, and CLI utility can implement a very simple logic to fetch ID and names mappings from API. This would also allow CLI to perform some interesting resolve mechanism, such as:
- 1 -> ID
- "1" -> name
- kind/breaking -> name
- "kind/breaking" -> name

What do you say @6543, wouldn't this be better abstracted from main Gitea code?

**Comment by 6543** [comment]
@CirnoT "resolving" the milestone by name was called "low efficient" so I go this way around
and indeed will this end up in one api call instead of 2

**Comment by CirnoT** [comment]
It might be easiest, but are we really going to support this for every single API that takes ID?

**Comment by 6543** [comment]
@CirnoT otherwise we have to do this on client side:
```go
// ResolveMilestoneIDByName is a fallback method to find milestone id by name
func (c *Client) ResolveMilestoneIDByName(owner, repo, name string) (int64, error) {
	i := 0
	for {
		i++
		miles, err := c.ListRepoMilestones(owner, repo, ListMilestoneOption{
			ListOptions: ListOptions{
				Page: i,
			},
			State: "all",
			Name: name,
		})
		if err != nil {
			return 0, err
		}
		if len(miles) == 0 {
			return 0, fmt.Errorf("milestone '%s' do not exist", name)
		}
		for _, m := range miles {
			if strings.ToLower(strings.TrimSpace(m.Title)) == strings.ToLower(strings.TrimSpace(name)) {
				return m.ID, nil
			}
		}
	}
}
```

**Comment by 6543** [comment]
> It might be easiest, but are we really going to support this for every single API that takes ID?
@CirnoT only if it realy make sence (to lower api requests)

**Review by lafriks — APPROVED** [review]

**Comment by lafriks** [comment]
:rocket:

**Pull request merged**